tools/rpkg/src/relational.cpp

```diff
@@ -188,6 +188,63 @@ external_pointer<T> make_external(const string &rclass, Args &&...args) {
 	return make_external<RelationWrapper>("duckdb_relation", res);
 }
 
+[[cpp11::register]] SEXP rapi_rel_left_join(duckdb::rel_extptr_t left, duckdb::rel_extptr_t right, list conds) {
+	unique_ptr<ParsedExpression> cond;
+
+	if (conds.size() == 0) { // nop
+		stop("rel_left_join needs conditions");
+	} else if (conds.size() == 1) {
+		cond = ((expr_extptr_t)conds[0])->Copy();
+	} else {
+		vector<unique_ptr<ParsedExpression>> cond_args;
+		for (expr_extptr_t expr : conds) {
+			cond_args.push_back(expr->Copy());
+		}
+		cond = make_unique<ConjunctionExpression>(ExpressionType::CONJUNCTION_AND, move(cond_args));
+	}
+
+	auto res = std::make_shared<JoinRelation>(left->rel, right->rel, move(cond), JoinType::LEFT);
```

Comment:
maybe it would be nicer to have a single rap_rel_join operator that takes a case-insensitive string(?) argument with the type defaulting to "inner"? There is some duplication here.

Reply:
@krlmlr any preference?

Reply:
For the caller, it doesn't matter. From the implementation side, it would seem that an extra argument would DRY things up in three or four places.

To avoid a breaking change, we can do

```r
rel_inner_join <- function(...) {
  rel_join(..., "inner")
```

with a suitable replacement for `...` .

What about cross and filtering (semi+anti) joins?

Reply:
I'm working on semi and anti joins, but they require a bit more attention since they use a combination of mark joins and filtering. I've been trying different things, but it's been proving difficult, I think it requires some modifications to the relational api